### PR TITLE
Add google cloud storage (gcs) support

### PR DESCRIPTION
**Pull request by ecopty**
This PR includes:

- Minimal GCSFileSystem
- GCSReadFile, GCSWriteFile
- Extends CircleCI to install/run the GCSFileSystem adapter. Includes a new job  linux-gcs-build as well as setup_adapters.sh script to download google-cloud-cpp storage and its dependencies.
- Basic testing of GCSUtils
- tests of the GCS file system read/write functionality using gcs emulator

### CODE REVIEW

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+  gcs::Client* client_;
+  std::string bucket_;
+  std::string key_;
+  int64_t size_{-1};
+  bool closed_{false};
+};
+} // namespace
+
+
+///////////////////
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:208]: remove?

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+class GCSConfig {
+ public:
+  GCSConfig(const Config* config) : config_(config)
+  {
+    std::string cred = config_->get("credentials", std::string(""));
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:213]: Should we name this `hive.gcs.credentials`?

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    if (!cred.empty())
+    {
+      credentials_ = gc::MakeServiceAccountCredentials(cred);
+    }
+  }
+
+  std::string endpointOverride() const {
+    return config_->get("endpoint_override", std::string(""));
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:221]: `hive.gcs.endpoint`?

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+  }
+
+  std::string scheme() const {
+
+    return config_->get("scheme", std::string("https"));
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:226]: `hive.gcs.scheme`?

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#include "gtest/gtest.h"
+
+using namespace facebook::velox;
+
+constexpr int kOneMB = 1 << 20;
+
+class GCSFileSystemTest : public testing::Test {
+ protected:
+  static void SetUpTestSuite() {
+
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp:32]: remove empty line.

**Review by majetideepak — CHANGES_REQUESTED** [review]
I found the discussion here https://issues.apache.org/jira/browse/ARROW-1231 helpful.
@ecopty It will be nice to include the emulator along with the tests in this PR for completeness. You can probably keep it as a separate commit to help simplify the review.

File: scripts/setup-adapters.sh [hunk 6]
```diff
@@ -16,6 +16,114 @@
 # Propagate errors and improve debugging.
 set -eufx -o pipefail
 
+function install_gcs-sdk-cpp {
+  #install google-cloud-cpp dependencies (ubuntu 20.04)
+  echo " Installing apt dependencies..."
+  export DEBIAN_FRONTEND=noninteractive
```
[the remark this answers is not in the document]
Comment (ecopty) [reply scripts/setup-adapters.sh:22]: > I found the discussion here https://issues.apache.org/jira/browse/ARROW-1231 helpful. @ecopty It will be nice to include the emulator along with the tests in this PR for completeness. You can probably keep it as a separate commit to help simplify the review.

sure @majetideepak, thank you for your reference, I was planning to add testing as a next step, but needed to do the research for emulator first (thanks!), in the meantime I appreciate your code review comments and I will attend to them.

File: scripts/setup-adapters.sh [hunk 7]
```diff
@@ -16,6 +16,114 @@
...
+  sudo apt-get update && \
+  sudo apt-get --no-install-recommends install -y apt-transport-https apt-utils \
+          automake build-essential ccache cmake ca-certificates curl git \
+          gcc g++ libc-ares-dev libc-ares2 libcurl4-openssl-dev libre2-dev \
+          libssl-dev m4 make pkg-config tar wget zlib1g-dev
+
+  #install abseil
```
[the remark this answers is not in the document]
Comment (ecopty) [reply scripts/setup-adapters.sh:29]: yep this whole part of the script is based on the info in the link you reference, did you mean for me to add the link there as documentation?

[the remark this answers is not in the document]
Comment (majetideepak) [reply scripts/setup-adapters.sh:29]: Adding a link in the setup helps for future reference.

**Comment by ecopty** [comment]
@majetideepak @kgpai @Yuhta 
Hi all, I have have added some tests, fixed issues raised in the previous code review. I am working on adding my company's CLA - this should be done soon. I would appreciate your review so I can close any issues while CLA requirement is being resolved.

Thanks

**Comment by majetideepak** [comment]
@ecopty Thanks for fixing the code review comments and adding tests. I will make another pass today.

File: .circleci/config.yml [hunk 8]
```diff
@@ -370,6 +371,58 @@ jobs:
             cd _build/release && ctest -j 16 -VV --output-on-failure
           no_output_timeout: 1h
 
+  linux-gcs-build:
```
Comment (majetideepak) [inline .circleci/config.yml:374]: FYI that https://github.com/facebookincubator/velox/pull/1427 will likely land before this. We should then extend `linux-parquet-hive-storage-adapters-build` run to test CGS as well. This works for now.

File: scripts/setup-adapters.sh [hunk 9]
```diff
@@ -20,6 +20,111 @@ SCRIPTDIR=$(dirname "${BASH_SOURCE[0]}")
...
+        -DBUILD_SHARED_LIBS=yes \
+        -DCMAKE_CXX_STANDARD=11 \
+        -H. -Bcmake-out && \
+      cmake --build cmake-out -- -j ${NCPU:-4} && \
+  $sudocmd cmake --build cmake-out --target install -- -j ${NCPU:-4} && \
+  $sudocmd ldconfig
+
+
+  # install protobuf
+  echo " Installing protobuf..."
```
Comment (majetideepak) [inline scripts/setup-adapters.sh:49]: `protobuf` is already a dependency for Velox. The minimum is `3.0.0`. The documentation of gcs says the minimum needed is `3.21.1`. Is that accurate? Looks like even `3.20.1` works. Two protobuf installs to default locations might lead to issues.
What do you think of installing all the artifacts related to `gcs` to a folder say `gcs_artifacts` and then point CMake to this folder?

File: scripts/setup-adapters.sh [hunk 10]
```diff
@@ -31,7 +136,41 @@ function install_aws-sdk-cpp {
...
+      shift # past argument
+      ;;
+    *)
+      echo "ERROR: Unknown option $1! will be ignored!"
+      shift
+      ;;
+  esac
+done
+
+if [ $install_gcs -eq 1 ]; then
```
Comment (majetideepak) [inline scripts/setup-adapters.sh:163]: Thanks for this change! Once the HDFS support lands, this refactor will help.

File: scripts/setup-ubuntu.sh [hunk 11]
```diff
@@ -86,6 +86,31 @@ function install_velox_deps {
   run_and_time install_folly
 }
 
+function install_gcs_deps {
```
Comment (majetideepak) [inline scripts/setup-ubuntu.sh:89]: Is this used anywhere? I guess you added `setup-gcs.sh` instead.

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+}
+
+std::string GCSFileSystem::name() const {
+  return "GCS";
+}
+
+folly::once_flag GCSInstantiationFlag;
+static std::function<std::shared_ptr<FileSystem>(std::shared_ptr<const Config>)>
+    filesystemGenerator = [](std::shared_ptr<const Config> properties) {
+      // Only one instance of GCSFileSystem is supported for now (follow S3 for
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:321]: S3 API is supported by at least 2 storage systems (Minio, AWS S3). So we might need to support two instances in the future.
However, GCS is only supported by Google, correct? Will we ever need two instances?

**Review by majetideepak — COMMENTED** [review]
@ecopty the changes look great! Have some minor comments.

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp (hunk 12 above)
Comment (majetideepak) [reply velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:321]: I guess we will need multiple instances for different credentials.

File: scripts/setup-ubuntu.sh (hunk 11 above)
Comment (ecopty) [reply scripts/setup-ubuntu.sh]: removed

File: scripts/setup-adapters.sh (hunk 9 above)
Comment (ecopty) [reply scripts/setup-adapters.sh:52]: @majetideepak I rebased my PR after the HDFS change and extended linux-parquet-hive-storage-adapters-build to include GCS as well. As for protobuf, I haven't seen any issue especially sinec 3.20.1 doesnt conflict with 3.0.0 so how about I uninstall (apt remove protobuf) before I install the 3.20.1

(I checked the 3.0.0 is not working for gcs)

Comment (majetideepak) [reply scripts/setup-adapters.sh:52]: @ecopty I don't think uninstall is an option since there could be other external dependencies. Since the adapter dependencies are installed in the `DEPENDENCY_DIR`, it should be okay.
Thanks for all the changes. I will make a final pass tomorrow.

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+
+#include <fmt/format.h>
+#include <glog/logging.h>
+#include <memory>
+#include <stdexcept>
+
+#include <google/cloud/storage/client.h>
+
+namespace facebook::velox {
+namespace filesystems {
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:30]: `S3ReadFile` is not part of the `filesystems` namespace in S3. Can we follow the same here for `GCSReadFile` and `GCSWriteFile` to be consistent?

File: velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+ private:
+  gcs::ObjectWriteStream stream_;
+  gcs::Client* client_;
+  std::string bucket_;
+  std::string key_;
+  int64_t size_{-1};
+  bool closed_{false};
+};
+} // namespace
+
```
Comment (majetideepak) [inline velox/connectors/hive/storage_adapters/gcs/GCSFileSystem.cpp:204]: `filesystems` namespace can begin here

**Review by majetideepak — COMMENTED** [review]
@ecopty changes look good to me. Have a couple more comments.

**Comment by majetideepak** [comment]
@ecopty can you address the last couple of comments. This should be ready for merge.

File: velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp [hunk 15]
```diff
@@ -346,19 +346,14 @@ TEST_F(GCSFileSystemTest, credentialsConfig) {
...
   } catch (VeloxException const& err) {
-    EXPECT_EQ(
-        err.message(),
-        std::string(
-            "Failed to get metadata for GCS object due to: Path:'gs://test1-gcs/test-object-name', "
-            "SDK Error Type:, GCS Status Code:Unknown error,  Message:"
-            "'Permanent error in GetObjectMetadata: Invalid "
-            "ServiceAccountCredentials,parsing failed on data loaded from memory'"));
+    EXPECT_THAT(err.message(), testing::HasSubstr("gs://test1-gcs/test-object-name"));
+    EXPECT_THAT(err.message(), testing::HasSubstr("Invalid ServiceAccountCredentials"));
```
Comment (bgmerrell) [inline velox/connectors/hive/storage_adapters/gcs/tests/GCSFileSystemTest.cpp:357]: @ecopty Does this change seem OK? (see the commit message for details)

**Comment by bgmerrell** [comment]
> @ecopty can you address the last couple of comments? This should be ready for merge after that.

Sorry for the delay with this. @ecopty had her sabbatical at intel and now will be leaving her team. I've been tasked with getting this PR accross the finish line. I don't have a lot of context or knowledge of this project, but I think I've addressed the review comments. Please take a look. I do see that the MacOS build is failing; please let me know if it's something wrong with this PR. Thank you!

**Comment by bgmerrell** [comment]
@majetideepak Hi, I just wanted to ping you directly on this to make sure you saw the updates :) thanks!

File: scripts/install_gcs_testbench.sh [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+
+if [ "$#" -ne 1 ]; then
+  echo "Usage: $0 <storage-testbench version>"
+  exit 1
+fi
```
Comment (Fiooodooor) [inline scripts/install_gcs_testbench.sh:21]: This from my perspective is at least unnecessary - this script is called exactly once and as build-time versions should be hardcoded due to dependencies and possible issues. 
```suggestion
```

File: scripts/setup-adapters.sh [hunk 17]
```diff
@@ -23,6 +23,111 @@ SCRIPTDIR=$(dirname "${BASH_SOURCE[0]}")
...
+      sed -i 's/^#define ABSL_OPTION_USE_\(.*\) 2/#define ABSL_OPTION_USE_\1 0/' "absl/base/options.h" && \
+      cmake \
+        -DCMAKE_BUILD_TYPE=Release \
+        -DBUILD_TESTING=OFF \
+        -DBUILD_SHARED_LIBS=yes \
+        -DCMAKE_CXX_STANDARD=11 \
+        -H. -Bcmake-out && \
+      cmake --build cmake-out -- -j ${NCPU:-4} && \
+  $sudocmd cmake --build cmake-out --target install -- -j ${NCPU:-4} && \
+  $sudocmd ldconfig
```
Comment (Mionsz) [inline scripts/setup-adapters.sh:48]: This is a really bad idea - part of files will be created as current user as an owner and part as root owner. There should be one check on script startup:

```bash
[ $(id -u) = 0 ] || echo 'Try again as root user.' && exit 1
```

File: scripts/setup-adapters.sh [hunk 18]
```diff
@@ -23,6 +23,111 @@ SCRIPTDIR=$(dirname "${BASH_SOURCE[0]}")
 source $SCRIPTDIR/setup-helper-functions.sh
 DEPENDENCY_DIR=${DEPENDENCY_DIR:-$(pwd)}
```
Comment (Mionsz) [inline scripts/setup-adapters.sh:25]:
```bash
function github_wget_and_untar {
  local URL=$1
  local DIR=$2
  mkdir -p "${DIR}"
  wget -q --max-redirect 3 -O - "https://github.com/${URL}" | tar -xz -C "${DIR}" --strip-components=1
  cd "${DIR}"
}
```

File: scripts/setup-adapters.sh [hunk 19]
```diff
@@ -23,6 +23,111 @@ SCRIPTDIR=$(dirname "${BASH_SOURCE[0]}")
...
+    root) echo 'Running as root' ; sudocmd="";;
+       *) echo 'Running as other user than root' ; sudocmd="sudo";;
+  esac
+  # install gcs dependencies
+  # https://github.com/googleapis/google-cloud-cpp/blob/main/doc/packaging.md#required-libraries
+  # install abseil
+  echo " Installing abseil..."
+  mkdir -p ${DEPENDENCY_DIR}/abseil-cpp && cd ${DEPENDENCY_DIR}/abseil-cpp
+  curl -sSL https://github.com/abseil/abseil-cpp/archive/20211102.0.tar.gz | \
+      tar -xzf - --strip-components=1 && \
```
Comment (Mionsz) [inline scripts/setup-adapters.sh:38]: For each and every of below steps it would go then:
`github_wget_and_untar 'abseil/abseil-cpp/archive/20211102.0.tar.gz' 'abseil-cpp'`

File: scripts/setup-adapters.sh [hunk 20]
```diff
@@ -23,6 +23,111 @@ SCRIPTDIR=$(dirname "${BASH_SOURCE[0]}")
...
+        -DBUILD_TESTING=OFF \
+        -DJSON_BuildTests=OFF \
+        -H. -Bcmake-out/nlohmann/json && \
+  $sudocmd cmake --build cmake-out/nlohmann/json --target install -- -j ${NCPU:-4} && \
+  $sudocmd ldconfig
+  echo " Clone, compile and install google-cloud-cpp..."
+  $sudocmd rm -rf ${DEPENDENCY_DIR}/google-cloud-cpp
+  # clone and compile the main project
+  git clone https://github.com/googleapis/google-cloud-cpp.git ${DEPENDENCY_DIR}/google-cloud-cpp
+  cd ${DEPENDENCY_DIR}/google-cloud-cpp
```
Comment (Mionsz) [inline scripts/setup-adapters.sh:118]: So in circle you have been using `~/adapter-deps/google-cloud-cpp-installed` and here the path is different?

File: scripts/setup-adapters.sh (hunk 17 above)
Comment (Mionsz) [inline scripts/setup-adapters.sh:48]: For each and every of below steps it would go then:
```suggestion
      sed -i 's/^#define ABSL_OPTION_USE_\(.*\) 2/#define ABSL_OPTION_USE_\1 0/' "absl/base/options.h"
      cmake_install  -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=yes -DCMAKE_INSTALL_PREFIX="${DEPENDENCY_DIR}/install" -DCMAKE_INSTALL_PREFIX="${DEPENDENCY_DIR}/install"
      ldconfig && \
      cd ..
```

File: velox/connectors/hive/storage_adapters/gcs/GCSUtil.h [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+  auto firstSep = path.find_first_of(kSep);
+  bucket = path.substr(0, firstSep);
+  key = path.substr(firstSep + 1);
+}
+inline std::string gcsURI(const std::string& bucket) {
+  return std::string(kGCSScheme) + bucket;
+}
+
+inline std::string gcsURI(const std::string& bucket, const std::string& key) {
+  return gcsURI(bucket) + "/" + key;
```
Comment (Mionsz) [inline velox/connectors/hive/storage_adapters/gcs/GCSUtil.h:48]: Question: so if there is a kSep introduced on line 24, shouldn't it be used here?
```suggestion
  return gcsURI(bucket) + std::string(kSep) + key;
```
And the same goes for S3 implementation, but this would be separate PR.

**Comment by obedmr** [comment]
@ecopty we (@tigrux and I) are joining this effort of adding support of gcs to Velox, would it be easy for you to grant us access in your branch? this in order to keep the advantage of all the history that is recorded on this PR. Other option can be that we fork the project, add your changes (you name would be set as the first contributor) and we may need to open a new PR. Please feel free to reach out for any further clarifications.

@majetideepak  and @Mionsz, we're just joining the effort, it will take us a few weeks to catch up with the project and changes but we'll come back with updates. Thanks for the review.
